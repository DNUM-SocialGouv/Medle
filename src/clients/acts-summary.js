import { ACTS_SUMMARY_ENDPOINT, API_URL } from "../config"
import { handleAPIResponse2 } from "../utils/errors"

export const findSummaryByHospital = async ({  hospitalId, headers }) => {
    const response = await fetch(`${API_URL}${ACTS_SUMMARY_ENDPOINT}?hospitals=${hospitalId}`, {
      headers: { ...headers, "Content-Type": "application/json" },
    })
  
    const { elements } = await handleAPIResponse2(response)
    return elements
  }

  export const fetchExport = async (selectedYear, hospitalId, headers) => {
    saveAs(`${API_URL}${ACTS_SUMMARY_ENDPOINT}/export?hospitals=${hospitalId}&year=${selectedYear}`, { ...headers })
  }
