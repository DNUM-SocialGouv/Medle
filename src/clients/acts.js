import { saveAs } from "file-saver"
import fetch from "isomorphic-unfetch"

import { ACTS_ENDPOINT, API_URL } from "../config"
import { handleAPIResponse } from "../utils/errors"
import { METHOD_DELETE, METHOD_POST, METHOD_PUT } from "../utils/http"

export const findAct = async ({ id, headers }) => {
  const response = await fetch(API_URL + ACTS_ENDPOINT + "/" + id, { headers })
  return handleAPIResponse(response)
}

export const searchActsByKey = async ({ key, value, headers }) => {
  const response = await fetch(`${API_URL}${ACTS_ENDPOINT}?${key}=${value}`, {
    headers: { ...headers, "Content-Type": "application/json" },
  })

  const { elements } = await handleAPIResponse(response)
  return elements
}

const buildQueryParams = (params) => {
  const { search, startDate, endDate, hospitals, profiles, asker, requestedPage } = params

  const arr = []
  if (search) {
    arr.push(`fuzzy=${search}`)
  }
  if (startDate) {
    arr.push(`startDate=${startDate}`)
  }
  if (endDate) {
    arr.push(`endDate=${endDate}`)
  }
  if (hospitals?.length) {
    arr.push(`hospitals=${hospitals.map((elt) => elt.value).join(",")}`)
  }
  if (profiles?.length) {
    arr.push(`profiles=${profiles.map((elt) => elt.value).join(",")}`)
  }
  if (asker?.value) {
    arr.push(`asker=${asker?.value}`)
  }
  if (requestedPage) {
    arr.push(`requestedPage=${requestedPage}`)
  }
  return arr.length ? "?" + arr.join("&") : ""
}

export const searchActs = async (params) => {
  const { headers } = params

  const queryParams = buildQueryParams(params)

  const response = await fetch(`${API_URL}${ACTS_ENDPOINT}${queryParams}`, { headers })

  return handleAPIResponse(response)
}

export const createAct = async ({ act, headers }) => {
  const response = await fetch(API_URL + ACTS_ENDPOINT, {
    method: METHOD_POST,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(act),
  })
  return handleAPIResponse(response)
}

export const updateAct = async ({ act, headers }) => {
  const response = await fetch(API_URL + ACTS_ENDPOINT + "/" + act.id, {
    method: METHOD_PUT,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(act),
  })
  return handleAPIResponse(response)
}

export const deleteAct = async ({ id, headers }) => {
  const response = await fetch(`${API_URL}${ACTS_ENDPOINT}/${id}`, { method: METHOD_DELETE, headers })
  return handleAPIResponse(response)
}

export const fetchExport = async (params) => {
  const { headers } = params

  const queryParams = buildQueryParams(params)

  saveAs(`${API_URL}${ACTS_ENDPOINT}/export${queryParams}`, { headers })
}
