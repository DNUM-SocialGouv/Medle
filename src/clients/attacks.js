import fetch from "isomorphic-unfetch"

import { API_URL, ATTACKS_ENDPOINT } from "../config"
import { handleAPIResponse2 } from "../utils/errors"
import { METHOD_DELETE, METHOD_POST, METHOD_PUT } from "../utils/http"

export const searchAttacksFuzzy = async ({ search, requestedPage, headers }) => {
  const arr = []
  if (search) {
    arr.push(`fuzzy=${search}`)
  }
  if (requestedPage) {
    arr.push(`requestedPage=${requestedPage}`)
  }
  const bonus = arr.length ? "?" + arr.join("&") : ""

  const response = await fetch(`${API_URL}${ATTACKS_ENDPOINT}${bonus}`, { headers })
  return handleAPIResponse2(response)
}

export const findAttack = async ({ id, headers }) => {
  const response = await fetch(`${API_URL}${ATTACKS_ENDPOINT}/${id}`, { headers })
  return handleAPIResponse2(response)
}

export const findAllAttacks = async ({ headers } = {}) => {
  const response = await fetch(API_URL + ATTACKS_ENDPOINT, { headers })
  return handleAPIResponse2(response)
}

export const createAttack = async ({ attack, headers }) => {
  const response = await fetch(`${API_URL}${ATTACKS_ENDPOINT}`, {
    body: JSON.stringify(attack),
    headers: { ...headers, "Content-Type": "application/json" },
    method: METHOD_POST,
  })

  return handleAPIResponse2(response)
}

export const updateAttack = async ({ attack, headers }) => {
  const response = await fetch(`${API_URL}${ATTACKS_ENDPOINT}/${attack.id}`, {
    body: JSON.stringify(attack),
    headers: { ...headers, "Content-Type": "application/json" },
    method: METHOD_PUT,
  })
  return handleAPIResponse2(response)
}

export const deleteAttack = async ({ id, headers }) => {
  const response = await fetch(`${API_URL}${ATTACKS_ENDPOINT}/${id}`, { headers, method: METHOD_DELETE })
  return handleAPIResponse2(response)
}
