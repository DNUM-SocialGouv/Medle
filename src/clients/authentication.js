import fetch from "isomorphic-unfetch"

import { API_URL, LOGIN_ENDPOINT } from "../config"
import { handleAPIResponse2 } from "../utils/errors"
import { METHOD_POST } from "../utils/http"

export const authenticate = async (email, password) => {
  const response = await fetch(API_URL + LOGIN_ENDPOINT, {
    method: METHOD_POST,
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email, password }),
  })

  const tokenCookie = response.headers.get("Set-Cookie") // may be null in CSR mode or because of CORS or secure cookie. OK in SSR and good for us for testing purpose.
  
  const { user, token } = await handleAPIResponse2(response)
  
  if (!user) {
    throw new Error("Authentication failed")
  }
  
  return {
    headers: {
      cookie: tokenCookie,
    },
    user,
    token
  }
}
