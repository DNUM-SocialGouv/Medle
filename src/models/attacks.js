import * as yup from "yup"

import * as common from "./common"

const JStoDBKeys = {
  id: "id",
  name: "name",
  year: "year",
}

const schema = yup.object().shape({
  id: yup.number().positive().integer().nullable(),
  name: yup.string(),
  year: yup.number().positive().integer(),
})

export const { transform, transformAll, untransform, validate } = common.build({ JStoDBKeys, schema })
