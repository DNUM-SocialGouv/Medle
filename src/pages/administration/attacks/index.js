import AddIcon from "@material-ui/icons/Add"
import Link from "next/link"
import { PropTypes } from "prop-types"
import React, { useState } from "react"
import { Alert, Col, Container, Form, FormGroup, Input, Spinner, Table } from "reactstrap"

import { searchAttacksFuzzy } from "../../../clients/attacks"
import { SearchButton } from "../../../components/form/SearchButton"
import Layout from "../../../components/Layout"
import Pagination from "../../../components/Pagination"
import { Title1 } from "../../../components/StyledComponents"
import { usePaginatedData } from "../../../hooks/usePaginatedData"
import { buildAuthHeaders, redirectIfUnauthorized, withAuthentication } from "../../../utils/auth"
import { logError } from "../../../utils/logger"
import { ADMIN } from "../../../utils/roles"

const AdminAttackPage = ({ paginatedData: initialPaginatedData, currentUser }) => {
  const [search, setSearch] = useState("")
  const [paginatedData, error, loading, fetchPage] = usePaginatedData(searchAttacksFuzzy, initialPaginatedData)

  const onChange = (e) => {
    setSearch(e.target.value)
  }

  const onSubmit = async (e) => {
    e.preventDefault()
    fetchPage({ search })(0)
  }

  return (
    <Layout page="attacks" currentUser={currentUser} admin={true}>
      <Container
        style={{ maxWidth: 980, minWidth: 740 }}
        className="mt-5 mb-5 d-flex justify-content-between align-items-baseline"
      >
        <Title1 className="">{"Administration des attentats"}</Title1>
        <Link href="/administration/attacks/[id]" as={`/administration/attacks/new`}>
          <a>
            <SearchButton className="btn-outline-primary">
              <AddIcon />
              &nbsp; Ajouter
            </SearchButton>
          </a>
        </Link>
      </Container>

      <Container style={{ maxWidth: 980, minWidth: 740 }}>
        <Form onSubmit={onSubmit}>
          <FormGroup row inline className="mb-4 justify-content-center">
            <Col className="flex-grow-1">
              <Input
                type="text"
                name="es"
                id="es"
                placeholder="Rechercher un attentat par son nom ou son année"
                value={search}
                onChange={onChange}
                autoComplete="off"
              />
            </Col>
            <Col className="flex-grow-0">
              <SearchButton className="btn-primary" disabled={loading} onClick={onSubmit}>
                {loading ? <Spinner size="sm" color="light" data-testid="loading" /> : "Chercher"}
              </SearchButton>
            </Col>
          </FormGroup>
        </Form>
        {error && (
          <Alert color="danger" className="mb-4">
            {error}
          </Alert>
        )}

        {!error && !paginatedData.elements.length && <div className="text-center">{"Aucun attentat trouvé."}</div>}

        {!error && !!paginatedData.elements.length && (
          <>
            <Pagination data={paginatedData} fn={fetchPage(search)} />

            <Table responsive className="table-hover">
              <thead>
                <tr className="table-light">
                  <th>Nom</th>
                  <th>Année</th>
                </tr>
              </thead>
              <tbody>
                {paginatedData.elements.map((attack) => (
                  <Link key={attack.id} href="/administration/attacks/[id]" as={`/administration/attacks/${attack.id}`}>
                    <tr>
                      <td>
                        <b>{`${attack.name}`}</b>
                      </td>
                      <td>
                        <b>{attack.year}</b>
                      </td>
                      <td>
                        <Link href="/administration/attacks/[id]" as={`/administration/attacks/${attack.id}`}>
                          <a className="text-decoration-none">Voir</a>
                        </Link>
                      </td>
                    </tr>
                  </Link>
                ))}
              </tbody>
            </Table>
          </>
        )}
      </Container>
    </Layout>
  )
}

AdminAttackPage.getInitialProps = async (ctx) => {
  const headers = buildAuthHeaders(ctx)

  try {
    const paginatedData = await searchAttacksFuzzy({ headers })
    return { paginatedData }
  } catch (error) {
    logError("APP error", error)

    redirectIfUnauthorized(error, ctx)
  }
  return {}
}

AdminAttackPage.propTypes = {
  paginatedData: PropTypes.object,
  currentUser: PropTypes.object.isRequired,
}

export default withAuthentication(AdminAttackPage, ADMIN)
