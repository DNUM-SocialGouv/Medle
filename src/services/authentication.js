import knex from "../knex/knex"
import { transform } from "../models/users"
import { compareWithHash } from "../utils/bcrypt"
import { APIError } from "../utils/errors"
import { STATUS_400_BAD_REQUEST, STATUS_401_UNAUTHORIZED } from "../utils/http"
import { generateToken } from "../utils/jwt"

const validPassword = (password) => !!password?.length

export const authenticate = async (email, password) => {
  // normalize email
  email = email?.trim()
  // request verification
  if (!validPassword(password)) {
    throw new APIError({
      status: STATUS_400_BAD_REQUEST,
      message: "Incorrect password",
    })
  }

  // SQL query
  const [dbUser] = await knex("users")
    .leftJoin("hospitals", "users.hospital_id", "hospitals.id")
    .where(knex.raw(`lower(email) = ?`, email?.toLowerCase()))
    .whereNull("users.deleted_at")
    .select(
      "users.id",
      "users.first_name",
      "users.last_name",
      "users.email",
      "users.password",
      "users.role",
      "users.hospital_id",
      "hospitals.name as hospital_name",
      "hospitals.extra_data as hospital_extra_data",
      "users.scope"
    )

  if (dbUser && (await compareWithHash(password, dbUser.password))) {
    const user = transform(dbUser)
    const token = generateToken(user)
    return { user, token }
  } else {
    // Unauthorized path
    throw new APIError({
      status: STATUS_401_UNAUTHORIZED,
      message: "Erreur d'authentification",
    })
  }
}
