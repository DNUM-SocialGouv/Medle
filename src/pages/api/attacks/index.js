import Cors from "micro-cors"

import { STATUS_200_OK, METHOD_GET, METHOD_OPTIONS } from "../../../utils/http"
import knex from "../../../knex/knex"
import { sendAPIError, sendMethodNotAllowedError } from "../../../services/errorHelpers"
import { ACT_CONSULTATION } from "../../../utils/roles"
import { checkValidUserWithPrivilege } from "../../../utils/auth"

const handler = async (req, res) => {
   res.setHeader("Content-Type", "application/json")

   try {
      switch (req.method) {
         case METHOD_GET: {
            checkValidUserWithPrivilege(ACT_CONSULTATION, req, res)

            const attacks = await knex("attacks")
               .whereNull("deleted_at")
               .orderBy("name")
               .select("id", "name")

            return res.status(STATUS_200_OK).json(attacks)
         }
         default:
            return sendMethodNotAllowedError(res)
      }
   } catch (error) {
      sendAPIError(error, res)
   }
}

const cors = Cors({
   allowMethods: [METHOD_GET, METHOD_OPTIONS],
})

export default cors(handler)
