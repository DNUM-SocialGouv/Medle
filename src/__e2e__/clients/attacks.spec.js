import { authenticate } from "../../clients/authentication"
import { findAllAttacks } from "../../clients/attacks"

const headersActUserTours = () => authenticate("acte@tours.fr", "test")

describe("/attacks", () => {
  it("should return all attacks", async () => {
    const { headers } = await headersActUserTours()
    const attacks = await findAllAttacks({ headers })

    expect(attacks).toMatchSnapshot()
  })
})
