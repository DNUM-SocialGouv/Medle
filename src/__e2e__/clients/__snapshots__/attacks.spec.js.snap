// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/attacks should return all attacks 1`] = `
Array [
  Object {
    "id": 5,
    "name": "2020 Villejuif",
  },
  Object {
    "id": 4,
    "name": "2016 Nice",
  },
  Object {
    "id": 3,
    "name": "2015 Les terrasses Paris",
  },
  Object {
    "id": 2,
    "name": "2015 Hyper Cacher",
  },
  Object {
    "id": 7,
    "name": "2015 Charlie Hebdo",
  },
  Object {
    "id": 1,
    "name": "2015 Bataclan",
  },
  Object {
    "id": 6,
    "name": "2012 Ã‰cole Ozar Hatorah Toulouse",
  },
]
`;
